providers/extensible_schema.go

```diff
@@ -105,13 +117,45 @@ func (x *extensibleSchema) LookupField(resource string, field string) (*resource
 		return found, nil
 	}
 
-	x.loadAllSchemas()
-	x.refresh()
+	x.unsafeLoadAll()
+	x.unsafeRefresh()
 
 	return found, found.Fields[field]
 }
 
-func (x *extensibleSchema) Add(name string, schema *resources.Schema) {
+// ---------------------------- unsafe methods ----------------------------
+// |  Only use these calls inside of a lock.                              |
+// |  Do NOT lock the object during these calls.                          |
+// V  Do NOT call to locking methods (~ everything above this line).      V
+// ------------------------------------------------------------------------
+
+func (x *extensibleSchema) unsafeLoadAll() {
+	// If another goroutine started to load this before us, it will be locked until
+	// we complete to load everything and then it will be dumped into this
+	// position. At this point, if it has been loaded we can return safely, since
+	// we don't unlock until we are finished loading.
+	if x.lastRefreshed >= LastProviderInstall {
+		return
+	}
+	x.lastRefreshed = LastProviderInstall
+
+	providers, err := ListActive()
```

Comment:
are all calls to `ListActive` guaranteed to be synchronized through a single `extensibleSchema`. I ask because this call ends up modifying global state without taking any locks

Reply:
Not in scope of this PR, we can follow up on this. I don't see a problem at this time, but we can secure providers separately too